language/evm/move-to-yul/src/yul_functions.rs

```diff
@@ -174,6 +189,33 @@ StorageKey: "(type, word) -> key {
   key := keccak256(${SCRATCH1_LOC}, 33)
 }",
 
+// Make a base storage offset for a given type. The result has 255 bits width and can be passed into
+// $MakePtr(true, result) to create a pointer as it fits into 255 bits. This pointer can be
+// used to linearly address exclusive memory with an address space of 60 bits.
+//
+//  254                                      0
+//  cccccc..cccccctttttt..tttttiiiii..iiiiiioooooo..oooooooo
+//   category       type_hash     id           offset
+//      4              32         160           60
+//
+// The category indicates what kind of type storage this is, and determines how id
+// is interpreted. RESOURCE_STORAGE_CATEGORY indicates that id is a resource
+// address. LINKED_STORAGE_CATEGORY indicates that id is a handle for data linked
+// to from some other storage (for instance, a vector aggregated by a resource).
+// The type_hash identifies the type of the stored value. The id is any 20 byte
+// number which identifies an instance of this type (e.g. an address if this is a resource).
+MakeTypeStorageBase: "(category, type_hash, instance) -> offs {
+  offs := or(shl(type_hash, 224), shl(instance, 64))
+}",
+
+// Make a new base storage offset for linked storage. This creates a new handle
+// and the calls MakeTypeStorageBase.
+NewLinkedStorageBase: "(type_hash) -> offs {
+  let handle := mload(${LINKED_STORAGE_COUNTER_LOC})
+  mstore(${LINKED_STORAGE_COUNTER_LOC}, add(handle, 1))
```

Comment:
IIUC this counter is for allocating space in the persistent (global) storage right? Wondering why we store the counter in memory instead of storage?

Reply:
Good catch! This counter need to go to global storage.